go/vt/vtgate/evalengine/eval.go

```diff
@@ -0,0 +1,315 @@
+		switch {
+		case v.IsSigned():
+			ival, err := v.ToInt64()
+			return newEvalFloat(float64(ival)), err
+		case v.IsUnsigned():
+			uval, err := v.ToUint64()
+			return newEvalFloat(float64(uval)), err
+		case v.IsFloat() || v.Type() == sqltypes.Decimal:
+			fval, err := v.ToFloat64()
+			return newEvalFloat(fval), err
+		case v.IsText() || v.IsBinary():
+			return newEvalFloat(parseStringToFloat(v.RawStr())), nil
+		default:
+			return nil, vterrors.Errorf(vtrpcpb.Code_INTERNAL, "coercion should not try to coerce this value to a float: %v", v)
+		}
+
+	case typ == sqltypes.Decimal:
+		var dec decimal.Decimal
+		switch {
+		case v.IsIntegral() || v.Type() == sqltypes.Decimal:
+			var err error
+			dec, err = decimal.NewFromMySQL(v.Raw())
+			if err != nil {
+				return nil, err
+			}
+		case v.IsFloat():
+			fval, err := v.ToFloat64()
+			if err != nil {
+				return nil, err
+			}
+			dec = decimal.NewFromFloat(fval)
+		case v.IsText() || v.IsBinary():
+			fval := parseStringToFloat(v.RawStr())
+			dec = decimal.NewFromFloat(fval)
+		default:
+			return nil, vterrors.Errorf(vtrpcpb.Code_INTERNAL, "coercion should not try to coerce this value to a decimal: %v", v)
+		}
+		return &evalDecimal{dec: dec, length: -dec.Exponent()}, nil
+
+	case sqltypes.IsSigned(typ):
+		switch {
+		case v.IsSigned():
+			ival, err := v.ToInt64()
+			return newEvalInt64(ival), err
+		case v.IsUnsigned():
+			uval, err := v.ToUint64()
+			return newEvalInt64(int64(uval)), err
+		default:
+			return nil, vterrors.Errorf(vtrpcpb.Code_INTERNAL, "coercion should not try to coerce this value to a signed int: %v", v)
+		}
+
+	case sqltypes.IsUnsigned(typ):
+		switch {
+		case v.IsSigned():
+			ival, err := v.ToInt64()
+			return newEvalUint64(uint64(ival)), err
+		case v.IsUnsigned():
+			uval, err := v.ToUint64()
+			return newEvalUint64(uval), err
+		default:
+			return nil, vterrors.Errorf(vtrpcpb.Code_INTERNAL, "coercion should not try to coerce this value to a unsigned int: %v", v)
+		}
+
+	case sqltypes.IsText(typ) || sqltypes.IsBinary(typ):
+		switch {
+		case v.IsText() || v.IsBinary():
+			// TODO: collation
+			return newEvalRaw(v.Type(), v.Raw(), collationBinary), nil
+		default:
+			return nil, vterrors.Errorf(vtrpcpb.Code_INTERNAL, "coercion should not try to coerce this value to a text: %v", v)
+		}
+	}
+	return nil, vterrors.Errorf(vtrpcpb.Code_INTERNAL, "coercion should not try to coerce this value: %v", v)
+}
+
+func valueToEvalNumeric(v sqltypes.Value) (eval, error) {
+	switch {
+	case v.IsSigned():
+		ival, err := v.ToInt64()
+		if err != nil {
+			return nil, vterrors.Errorf(vtrpcpb.Code_INVALID_ARGUMENT, "%v", err)
+		}
+		return &evalInt64{ival}, nil
+	case v.IsUnsigned():
+		var uval uint64
+		uval, err := v.ToUint64()
+		if err != nil {
+			return nil, vterrors.Errorf(vtrpcpb.Code_INVALID_ARGUMENT, "%v", err)
+		}
+		return newEvalUint64(uval), nil
+	default:
+		uval, err := strconv.ParseUint(v.RawStr(), 10, 64)
+		if err == nil {
```

Comment:
The logic in `NewLiteralIntegralFromBytes` checks if it fits in an `int64`, do we need something similar here to be consistent with that? Or does that not matter?

Reply:
No, the `NewLiteralXXX` functions do extra checking to ensure the SQL Parser or people manually using the API do not overflow the numeric types. This is an internal API, however, and MySQL always silently handles these overflows.